[pytorch] fix CUDA_KERNEL_ASSERT macro for android build

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40151 [pytorch] fix CUDA_KERNEL_ASSERT macro for android build**

Summary:
For debug android build it throws the following error:
```
  In file included from src/pytorch/android/pytorch_android/src/main/cpp/pytorch_jni_common.cpp:9:
  In file included from src/pytorch/android/pytorch_android/src/main/cpp/pytorch_jni_common.h:2:
  In file included from ../../../../src/main/cpp/libtorch_include/armeabi-v7a/torch/csrc/api/include/torch/types.h:3:
  In file included from ../../../../src/main/cpp/libtorch_include/armeabi-v7a/ATen/ATen.h:5:
  In file included from ../../../../src/main/cpp/libtorch_include/armeabi-v7a/ATen/Context.h:4:
  In file included from ../../../../src/main/cpp/libtorch_include/armeabi-v7a/ATen/Tensor.h:3:
  In file included from ../../../../src/main/cpp/libtorch_include/armeabi-v7a/ATen/core/TensorBody.h:7:
  In file included from ../../../../src/main/cpp/libtorch_include/armeabi-v7a/c10/core/Scalar.h:13:
  ../../../../src/main/cpp/libtorch_include/armeabi-v7a/c10/util/TypeCast.h:157:22: error: use of undeclared identifier '__assert_fail'
  AT_FORALL_QINT_TYPES(DEFINE_UNCASTABLE)
                       ^
```

Seems `__assert_fail()` isn't available on Android by default - in NDEBUG mode it forward declares the function and CI passes.

But CUDA_KERNEL_ASSERT() shouldn't be relevant for mobile build at all and we already bypass `__APPLE__` so the easiest fix is to add `__ANDROID__`.

Differential Revision: [D22095562](https://our.internmc.facebook.com/intern/diff/D22095562)

ljk53 committed:
[pytorch] fix CUDA_KERNEL_ASSERT macro for android build

Summary:
For debug android build it throws the following error:
```
  In file included from src/pytorch/android/pytorch_android/src/main/cpp/pytorch_jni_common.cpp:9:
  In file included from src/pytorch/android/pytorch_android/src/main/cpp/pytorch_jni_common.h:2:
  In file included from ../../../../src/main/cpp/libtorch_include/armeabi-v7a/torch/csrc/api/include/torch/types.h:3:
  In file included from ../../../../src/main/cpp/libtorch_include/armeabi-v7a/ATen/ATen.h:5:
  In file included from ../../../../src/main/cpp/libtorch_include/armeabi-v7a/ATen/Context.h:4:
  In file included from ../../../../src/main/cpp/libtorch_include/armeabi-v7a/ATen/Tensor.h:3:
  In file included from ../../../../src/main/cpp/libtorch_include/armeabi-v7a/ATen/core/TensorBody.h:7:
  In file included from ../../../../src/main/cpp/libtorch_include/armeabi-v7a/c10/core/Scalar.h:13:
  ../../../../src/main/cpp/libtorch_include/armeabi-v7a/c10/util/TypeCast.h:157:22: error: use of undeclared identifier '__assert_fail'
  AT_FORALL_QINT_TYPES(DEFINE_UNCASTABLE)
                       ^
```

Seems `__assert_fail()` isn't available on Android - in NDEBUG mode the macro uses `_wassert()` which is available (thus CI passes).

But CUDA_KERNEL_ASSERT() shouldn't be relevant for mobile build at all and we already bypass `__APPLE__` so the easiest fix is to add `__ANDROID__`.

[ghstack-poisoned]

Baranowski commented on c10/macros/Macros.h:
```diff
@@ -212,7 +212,7 @@ constexpr uint32_t CUDA_THREADS_PER_BLOCK_FALLBACK = 256;
 // CUDA_KERNEL_ASSERT checks the assertion
 // even when NDEBUG is defined. This is useful for important assertions in CUDA
 // code that when building Release.
```
While you are here, could you please also fix the comment above your change? For example

> CUDA_KERNEL_ASSERT checks the assertion even when NDEBUG is defined. This is useful for important assertions in CUDA code that **would otherwise be suppressed** when building Release.

Baranowski reviewed (approved):
Thanks, LGTM

ljk53 committed:
Update on "[pytorch] fix CUDA_KERNEL_ASSERT macro for android build"


Summary:
For debug android build it throws the following error:
```
  In file included from src/pytorch/android/pytorch_android/src/main/cpp/pytorch_jni_common.cpp:9:
  In file included from src/pytorch/android/pytorch_android/src/main/cpp/pytorch_jni_common.h:2:
  In file included from ../../../../src/main/cpp/libtorch_include/armeabi-v7a/torch/csrc/api/include/torch/types.h:3:
  In file included from ../../../../src/main/cpp/libtorch_include/armeabi-v7a/ATen/ATen.h:5:
  In file included from ../../../../src/main/cpp/libtorch_include/armeabi-v7a/ATen/Context.h:4:
  In file included from ../../../../src/main/cpp/libtorch_include/armeabi-v7a/ATen/Tensor.h:3:
  In file included from ../../../../src/main/cpp/libtorch_include/armeabi-v7a/ATen/core/TensorBody.h:7:
  In file included from ../../../../src/main/cpp/libtorch_include/armeabi-v7a/c10/core/Scalar.h:13:
  ../../../../src/main/cpp/libtorch_include/armeabi-v7a/c10/util/TypeCast.h:157:22: error: use of undeclared identifier '__assert_fail'
  AT_FORALL_QINT_TYPES(DEFINE_UNCASTABLE)
                       ^
```

Seems `__assert_fail()` isn't available on Android - in NDEBUG mode the macro uses `_wassert()` which is available (thus CI passes).

But CUDA_KERNEL_ASSERT() shouldn't be relevant for mobile build at all and we already bypass `__APPLE__` so the easiest fix is to add `__ANDROID__`.

[ghstack-poisoned]

ljk53 commented:
Just to keep a record that @IvanKobzarev made a valid point that this assert can be shared between cuda & cpu sources: https://github.com/pytorch/pytorch/pull/39584#discussion_r441736759

We can redefine the assert to some other assert() available on Android if needed.
